sql: configure zone config for CREATE DATABASE

otan opened the pull request:
Release note (sql change): CREATE DATABASE ... WITH [PRIMARY] REGIONS
now modifies the zone configurations with the settings as set by the
multiregion configuration.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57244)
<!-- Reviewable:end -->

ajstorm reviewed (changes requested):
Great to see the first cut of zone configs!  A few comments below, and bringing in @aayushshah15 for a second opinion.

> Reviewed 5 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-:-MNUhKLz3-hQV9qhQD30:bjq1orx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @ajstorm, and @otan)

___
*[pkg/sql/region_util.go, line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNUNvnP6wlvz4uCRzor:-MNUNvnP6wlvz4uCRzos:bd2arph) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util.go#L82)):*
> ```Go
> }
> 
> func makeRequireZoneConstraintForRegion(r descpb.Region) zonepb.Constraint {
> ```

Nit: should this be makeRequiredZoneConstraintForRegion()?
___
*[pkg/sql/region_util.go, line 90 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNUQ9mMCAFWM3Cnas3L:-MNUQ9mMCAFWM3Cnas3M:b-1wsjyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util.go#L90)):*
> ```Go
> }
> 
> func genZoneConfigFromRegionConfigForDatabase(
> ```

Does the logic in here assume "SURVIVE REGION FAILURE"?  If so, we should make that clear from the comment.  Also, do we think that we're going to have separate functions for this case and SURVIVE ZONE?  If so, we might want to change the function name here to be more reflective of the case it's solving.
___
*[pkg/sql/region_util.go, line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNURAiJ6otLRhqeKig4:-MNURAiJ6otLRhqeKig5:brgxnb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util.go#L128)):*
> ```Go
> 		genZoneConfigFromRegionConfigForDatabase(regionConfig),
> 	)
> 	if err != nil {
> ```

Nit:  I'm not picky either way, but it's a bit strange that you check the err value after the call here, and in a compound if statement below.  Might read better if you're consistent within this function.
___
*[pkg/sql/region_util_test.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNUU_US5ZE47twNedgO:-MNUU_US5ZE47twNedgP:bgvrqjw) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util_test.go#L53)):*
> ```Go
> 				Regions:       []descpb.Region{"region_a"},
> 				PrimaryRegion: "region_a",
> 				SurvivalGoal:  descpb.SurvivalGoal_REGION_FAILURE,
> ```

Isn't this not allowed higher up in the stack?  We can't survive a region failure with only one region.
___
*[pkg/sql/region_util_test.go, line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNUSetAEw7pcclpsGNF:-MNUSetAEw7pcclpsGNG:b7cosd9) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util_test.go#L77)):*
> ```Go
> 					{Constraints: []zonepb.Constraint{{Type: zonepb.Constraint_REQUIRED, Key: "region", Value: "region_a"}}},
> 				},
> 				Constraints: []zonepb.ConstraintsConjunction{
> ```

I might be misunderstanding things, but I think the expected case is not correct.  With zone failures, and only two regions, I think you'll need at least 4 (but more likely 5, as we need to round to an odd number) replicas.  With the way you have it now, one of the 3 replicas will be in the non-primary region, therefore if one replica goes down, we'll be forced to incur cross-region latencies.  It's a bit of an edge case, but I think we'd want to ensure intra-region quorum for REGIONAL BY TABLE tables, even if one replica goes down.  @aayushshah15 thoughts?
___
*[pkg/sql/region_util_test.go, line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNUbXXfDJN-5pouInsg:-MNUbXXg4CEUnXczS9ZK:b-45twx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util_test.go#L88)):*
> ```Go
> 				Regions:       []descpb.Region{"region_b", "region_a"},
> 				PrimaryRegion: "region_a",
> 				SurvivalGoal:  descpb.SurvivalGoal_REGION_FAILURE,
> ```

This should also be prevented higher up.  No point in testing it down here.
___
*[pkg/sql/region_util_test.go, line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNUcDf94ml0jpYhjm-L:-MNUcDf94ml0jpYhjm-M:bgp5hra) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util_test.go#L109)):*
> ```Go
> 			},
> 			expected: &zonepb.ZoneConfig{
> 				NumReplicas: proto.Int32(3),
> ```

I think this one is incorrect too.  This needs to be 5 replicas to ensure that we get intra-region latencies for reads/writes.
___
*[pkg/sql/region_util_test.go, line 147 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNUcS4Z5LAiXo4fRVQv:-MNUcS4Z5LAiXo4fRVQw:b-iif2mj) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util_test.go#L147)):*
> ```Go
> 			},
> 			expected: &zonepb.ZoneConfig{
> 				NumReplicas: proto.Int32(4),
> ```

I think that this should be at least 5, but more reasonably 6 (or 7, if we want something odd) for the same reasons above.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57244)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @ajstorm)

___
*[pkg/sql/region_util.go, line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNUNvnP6wlvz4uCRzor:-MNUmbr0CB3RJ_z2YRaX:b-taqnoy) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util.go#L82)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Nit: should this be makeRequiredZoneConstraintForRegion()?
</blockquote></details>

Yep! Done.
___
*[pkg/sql/region_util.go, line 90 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNUQ9mMCAFWM3Cnas3L:-MNUmcvg7k90oSi3V3Ds:b-lpcm40) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util.go#L90)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Does the logic in here assume "SURVIVE REGION FAILURE"?  If so, we should make that clear from the comment.  Also, do we think that we're going to have separate functions for this case and SURVIVE ZONE?  If so, we might want to change the function name here to be more reflective of the case it's solving.
</blockquote></details>

Made it clear I am supporting aayush's playbook instructions initially, and he'll be changing these later.
___
*[pkg/sql/region_util.go, line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNURAiJ6otLRhqeKig4:-MNUmigj95I1yP353Vhq:b-3zpz35) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util.go#L128)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Nit:  I'm not picky either way, but it's a bit strange that you check the err value after the call here, and in a compound if statement below.  Might read better if you're consistent within this function.
</blockquote></details>

can't do that, because `if sql, err := ...` is scoped to the `if` clause. since we need `sql` is needed after the `if` statement, we can't do that :) 
___
*[pkg/sql/region_util_test.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNUU_US5ZE47twNedgO:-MNUn0m79seYN6-4WESf:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util_test.go#L53)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Isn't this not allowed higher up in the stack?  We can't survive a region failure with only one region.
</blockquote></details>

Removed.
___
*[pkg/sql/region_util_test.go, line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNUSetAEw7pcclpsGNF:-MNUn32Y08I4byLnDDIA:bw6inep) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util_test.go#L77)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

I might be misunderstanding things, but I think the expected case is not correct.  With zone failures, and only two regions, I think you'll need at least 4 (but more likely 5, as we need to round to an odd number) replicas.  With the way you have it now, one of the 3 replicas will be in the non-primary region, therefore if one replica goes down, we'll be forced to incur cross-region latencies.  It's a bit of an edge case, but I think we'd want to ensure intra-region quorum for REGIONAL BY TABLE tables, even if one replica goes down.  @aayushshah15 thoughts?
</blockquote></details>

Clarified as a comment so far.
___
*[pkg/sql/region_util_test.go, line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNUbXXfDJN-5pouInsg:-MNUn5tB5kC37bIvbmNf:ba3dfob) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util_test.go#L88)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

This should also be prevented higher up.  No point in testing it down here.
</blockquote></details>

Removed
___
*[pkg/sql/region_util_test.go, line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNUcDf94ml0jpYhjm-L:-MNUn7af6RRA7LfUBpDF:bm4odi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util_test.go#L109)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

I think this one is incorrect too.  This needs to be 5 replicas to ensure that we get intra-region latencies for reads/writes.
</blockquote></details>

Clarified with comment with understanding aayush will modify this.
___
*[pkg/sql/region_util_test.go, line 147 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNUcS4Z5LAiXo4fRVQv:-MNUnBJo9UHikchsraxR:bm4odi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util_test.go#L147)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

I think that this should be at least 5, but more reasonably 6 (or 7, if we want something odd) for the same reasons above.
</blockquote></details>

Clarified with comment with understanding aayush will modify this.


<!-- Sent from Reviewable.io -->

otan committed:
sql: configure zone config for CREATE DATABASE

Release note (sql change): CREATE DATABASE ... WITH [PRIMARY] REGIONS
now modifies the zone configurations with the settings as set by the
multiregion configuration.

ajstorm reviewed (approved):
Looks great.  Thanks for making those changes!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-:-MNZ4YERA21FIcqPQkeo:bjpzglm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @ajstorm)

___
*[pkg/sql/region_util.go, line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57244#-MNURAiJ6otLRhqeKig4:-MNZ4-1iEYMbW4uTEbuc:b-mkalkq) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a94917468b9696da14dc4dd7790c17d175b92b/pkg/sql/region_util.go#L128)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

can't do that, because `if sql, err := ...` is scoped to the `if` clause. since we need `sql` is needed after the `if` statement, we can't do that :) 
</blockquote></details>

ahh, didn't catch that sql was used later on.  I withdraw my comment :)


<!-- Sent from Reviewable.io -->

otan commented:
discussed with @aayushshah15 -- merging for now, he will modify anything at a later stage.

bors r=ajstorm
